refactor: Flutter v 3.7.0

M123-dev opened the pull request:
### What
Upgrading to flutter 3.7.0 including ci / cd migration


The upgrade itself was done in a few minutes with just some imports removed and renaming of a few parameters, but when then trying to run I got an unspecified null error in the language generation together with a lot of new warning inside of them, thats why I had to refactor them all.

- `one {}` was oftentimes overwritten by `=1{}` 
- `=3` and `=4` are not valid
- A underscore in placeholders are not allowed
- A closing `}` in every `plural_compare_x_products`

I'm open whether we should use 3.7.0 or maybe a somewhat older version for stability, that will just be a few lines in the workflows. The whole dart & localization changes will apply to versions >=3.3.0

M123-dev committed:
chore: Flutter upgrade

M123-dev committed:
fix: Local warnings

M123-dev committed:
fix: Locale

M123-dev committed:
fix: collection package

monsieurtanuki reviewed (commented):
Hi @M123-dev!
As you say, it's not certain that upgrading to 3.7.0 is a good idea. But staying in 3.0.5 is even a worse idea, so thank you for this PR!
That means we have to be especially careful once this PR is merged, and fully manually test the app.

Regarding the additional fixes you mentioned...
> * one {} was oftentimes overwritten by =1{}
> * =3 and =4 are not valid
> * A underscore in placeholders are not allowed
> * A closing } in every plural_compare_x_products

In each case, could you state if this impacts many labels, if it's still working and if there's a workaround?
For instance
* I expect the `one`/`1` to still work though it impacts a lot a labels
* I expect the `3`/`4` to impact only not that many labels in not that many languages

M123-dev commented:
Only the ones in the _en.arb are only really affected, as the other ones are just local copies who aren't even uploaded to the server. 

I just counted together

7x dublicate one
1x placeholder name change

\+ 1x missing } which is also affected by the dublicate one


The =3 and =4 aren't in the English file, so they have to be fixed in crowdin

So overall 8 labels which need to be retranslated after that. 
Honestly I didn't check which of the problems caused the error, it was a unspecified null error which even suggested opening a issue in flutter. I could try to revert some of the changes to avoid retranslation, though 8 are comparably not that much and eventually they should be fixed anyways.

monsieurtanuki commented:
@M123-dev As there are few potential issues, please check in each case that the translation still works.

> one {} was oftentimes overwritten by =1{}

Please have a look at the new correct syntax and remove the duplicates, so that we always have the same tag in the whole file, either `one` or `=1`
Current
`"plural_ago_weeks": "{count,plural, one {} =1{one week ago} other{{count} weeks ago}}",`
should probably become (even in flutter 3.0.5)
`"plural_ago_weeks": "{count,plural, =1{one week ago} other{{count} weeks ago}}",`

> =3 and =4 are not valid

Looks like the problem is only in `app_pl.arb`, for 4 labels. Please have a look at the correct syntax for 3 and 4. Worst case scenario: it's not possible, and we won't be able to spell correctly in Polish 4 labels for counts 3 and 4. That should not prevent us from upgrading flutter.

> A underscore in placeholders are not allowed

Please fix the only placeholder that seems to be impacted, in the code and in the .arb files

> A closing } in every plural_compare_x_products

`"plural_compare_x_products": "{count,plural, =1{Compare one Product} other{Compare {count} Products}",`
That makes sense to add a matching closing curly bracket, even in flutter 3.0.5.

M123-dev commented:
All of these problems are already fixed in the PR :D

And yes sadly there is no way to specify something for 3 or 4 or anything else then 0,1,2

monsieurtanuki commented:
> All of these problems are already fixed in the PR :D

Cool! 

> And yes sadly there is no way to specify something for 3 or 4 or anything else then 0,1,2

Not sure it's a problem as I checked in [reverso.net](https://www.reverso.net/):
| EN | PL |
| -- | -- |
| 1 hour ago | 1 godzinę temu |
| 2 hours ago | 2 godzin temu |
| 3 hours ago | 3 godzin temu |
| 4 hours ago | 4 godzin temu |
| 5 hours ago | 5 godzin temu |

The problem is that I get different results in https://translate.google.com/ (same results as in `app_pl.arb`).
Very strange.
I guess that should not stop us from upgrading.
We'll see if users react strongly with those small mistakes...

M123-dev committed:
Merge branch 'develop' into flutter-upgrade-3.7.0

M123-dev committed:
fix: Badge import

M123-dev committed:
autofix: TextTheme

monsieurtanuki commented on packages/smooth_app/lib/background/background_task_badge.dart:
```diff
@@ -1,4 +1,4 @@
-import 'package:badges/badges.dart';
+import 'package:badges/badges.dart' as badges;
```
Actually the idea would be instead to use the standard flutter `Badge` that is available after 3.0.5.
And get rid of `badges.dart` in `pubspec.yaml`, as commented.

M123-dev committed:
lots of use_build_context_synchronously

M123-dev committed:
Update extension_on_text_helper.dart

M123-dev commented:
For some reason, there appeared lots more problems, mostly about asynchronous context but some others as well. I would rather leave them as they are right now and fix them in follow-up PR's.

M123-dev committed:
use_build_context_synchronously

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/congrats.dart:
```diff
@@ -90,6 +90,7 @@ class CongratsWidget extends StatelessWidget {
                 ),
               );
               if (OpenFoodAPIConfiguration.globalUser != null) {
+                // ignore: use_build_context_synchronously
```
Please use `if(!mounted) { return; }` instead.
I think that does work on `StatelessWidget` too from 3.7.

monsieurtanuki commented on packages/smooth_app/lib/pages/preferences/user_preferences_account.dart:
```diff
@@ -318,7 +318,9 @@ class _UserPreferencesPageState extends State<UserPreferencesSection> {
           () async {
+            // ignore: use_build_context_synchronously
             if (await _confirmLogout(context) == true) {
+              // ignore: use_build_context_synchronously
```
For this guy, you can also start your code with `final NavigatorState navigatorState = Navigator.of(context)`, and then `navigatorState.pop()`.

monsieurtanuki reviewed (approved):
Hi @M123-dev!
I've already said incrementally what I had to on different commits (like `Badge` and `ignore`).
I won't review the 203 files again.
What's prevented you from merging days ago? Still some lint issues?

M123-dev committed:
format

M123-dev committed:
fix: locale not using the placeholder

This is not intended to be used in a long term, but it will allow us to migrate
c.f. https://github.com/flutter/flutter/issues/119557#issuecomment-1411088910

M123-dev commented:
Okay now we should be hopefully ready to merge, stumbled across the next error in the locale generation c.f. https://github.com/flutter/flutter/issues/119557#issuecomment-1411088910

M123-dev commented:
Thanks for your review @monsieurtanuki

monsieurtanuki commented:
@M123-dev Thank you for the PR, now we're on 3.7, right?
The bug you mentioned is reasonable: it's probably a translation mistake if the placeholder is not in all translations. The workaround (if needed and if not done already) would be to artificially add the placeholder in the problematic translations.
